vis_sink.py

```diff
@@ -0,0 +1,374 @@
+
+        # objects port processing
+        while self.objects_port.getPendingReads() != 0:
+            objects_bottle = self.objects_port.read(True)
+            if objects_bottle:
+                objects_bottles = list(
+                    map(yarp.Value.asList,
+                        list(
+                            map(objects_bottle.get,
+                                list(range(objects_bottle.size()))))))
+                for object_bottle in objects_bottles:
+                    object_id = object_bottle.get(0).asInt()
+                    object_prop_name = object_bottle.get(1).toString()
+                    if object_prop_name == 'type':
+                        object_prop_data = object_bottle.get(2).toString()
+                    else:
+                        object_prop_data = array(
+                            list(
+                                map(yarp.Value.asDouble,
+                                    list(
+                                        map(object_bottle.get,
+                                            list(range(
+                                                object_bottle.size()))[2:])))))
+                    # Properties are: type (sphere, frame, arrow, box), pose
+                    # (16 values), color (three values), scale (three values),
+                    # etc
+                    self.add_log({
+                        "id": object_id,
+                        "object_prop_name": object_prop_name,
+                        "object_prop_data": object_prop_data
+                    })
+
+    def get_hand_info(self, hand_port, side_str):
+        angles_field = list(range(3))
+        speeds_field = list(range(3, 6))
+        torques_field = list(range(6, 9))
+        THUMB_ANGLE_FIELD = 1
+        THUMB = 0
+        FIRST_FINGER = 1
+        MIDDLE_FINGER = 2
+        RING_FINGER = 3
+        PINKY_FINGER = 4
+        all_finger_list = [
+            THUMB, FIRST_FINGER, MIDDLE_FINGER, RING_FINGER, PINKY_FINGER
+        ]
+
+        bottle = hand_port.read()
+        # reading data from server (all fingers always)
+        for fingerbottle, finger in zip(
+                list(
+                    map(yarp.Value.asList,
+                        list(map(bottle.get, list(range(2, bottle.size())))))),
+                all_finger_list):
+            if finger == THUMB:
+                thumb_angle = bottle.get(
+                    THUMB_ANGLE_FIELD).asDouble()  # for thumb position
+                thumb_speed = 0.0  # TODO: Calculate thumb speed
+                self.add_log({
+                    "id":
+                    "Hand" + side_str,
+                    "finger":
+                    finger,
+                    "angles": ([thumb_angle] + list(
+                        map(yarp.Value.asDouble,
+                            list(map(fingerbottle.get, angles_field))))),
+                    "speeds": ([thumb_speed] + list(
```

Comment:
lists have an `extend()` method that is better suited for this task: [docs](https://docs.python.org/2/tutorial/datastructures.html)